### PR TITLE
[LLVMCPU] WIP: Add softmax ukernel for riscv64

### PR DESCRIPTION
**Pull request by ElEHsiang**
# Description
This patch adds softmax f32 ukernel for riscv64.
The softmax ukernel is lowering from linalg.softmax.

On 12x128x40960xf32 test case, the ukernel can improve the performance about 50% with single thread and about 70% with multithread. For the mobilebert f32 model, it can improve the performance about 30%.

The DecomposeSoftmaxPass is moved out from addCommonTargetExecutablePreprocessPass so the linalg.softmax op will not be decomposed.

# Discussion
1. Seems can't easily use `check_cxx_compiler_flag` to check riscv v extension, I enable v extension in `riscv.toolchain.cmake` currently.

2. Currently, we can't use the hwprobe in linux so I parse `/proc/cpuinfo` to check whether the CPU supports the v extension.
(Ps. the riscv hwprobe is available after 6.5)

3. The `LLVMCPULowerToUKernelsPass` is added to `addCPUDefaultPassPipeline`, is it better to put in another place?

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1749]: This will end up using the default pipeline by default and will result in performance regressions since no vectorization etc. kicks in.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
I'd strongly reconsider such a lowering. Softmax is pretty "large operation" to be passed to a  micro kernel. The benefits here are great, but I cant immediately follow what the cause for the improvements are. It seems like the goal here is to effectively take softmax out of the normal codegen flow and use a "library implementation" for softmax. 

In any case, the plumbing of this is effectively taking out softmax for all CPU backends from the normal codegen lowering path. Thats a pretty heavy hammer. I would drill down a little bit more to figure out why exactly the performance of the hand implemented kernel is better. For the existing MMT4d lowerings, the main justification is that getting the right instructions to be selected for the different data types is non-trivial.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (ElEHsiang) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1749]: Oh, It should use another pipeline, I'll change it.

**Comment by ElEHsiang** [comment]
This patch is use the "library implementation" for softmax on riscv64. It is a convenient way to improve the performance if the CPU backend already has an optimized kernel. I can do some comparisons between the optimized softmax and codegen softmax. Ideally we want the compiler generate the optimized code, but it seems not easy in the short term.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1749]: Thats not the issue though... this is changing the behavior for all CPU deployments (x86 and ARM)....

**Comment by MaheshRavishankar** [comment]
> This patch is use the "library implementation" for softmax on riscv64. It is a convenient way to improve the performance if the CPU backend already has an optimized kernel. I can do some comparisons between the optimized softmax and codegen softmax. Ideally we want the compiler generate the optimized code, but it seems not easy in the short term.

Softmax is not really lowered as a softmax directly. It is first broken up into several reductions and then vectorized. If you are looking for a library implementation for softmax, then microkernels are not the way to go. You are looking for some way of shipping a pre-baked library kernel in the compiler.... There are other approaches that have been developed to do library offload using the plugin mechanisms.. those are all pessimizing though... Ideally a microkernel is working on "small tiles", think something like L1 resident.

**Comment by rednoah91** [comment]
Hi @MaheshRavishankar, our first intention is to provide a library alternative for users to choose in the short term, and then as a reference to help the compiler to generate better code in the long term. As you mentioned, if a microkernel isn't a recommended approach, could you please guide us to the right mechanism?
Another question is: once we have a L1 tiles ukernel, how should we integrate it? Thanks.

File: runtime/src/iree/base/internal/cpu.c
Comment (ScottTodd) [inline runtime/src/iree/base/internal/cpu.c:389]: This riscv CPU initialization code looks useful to extract into a standalone PR. Splitting into multiple PRs helps keep the review discussions focused and moving, and it looks like the softmax bits are contentious.

**Comment by benvanik** [comment]
Agreed with Mahesh - we want to enable this kind of coarse library-call style kernel but don't generally want them in the core repository - but is still cool, and great work! This is a great example of using IREE to implement a specialization and I think after what Mahesh suggests (investigating why this is a benefit at the coarse level vs if it's just an instruction selection thing) if it still holds as a coarse implementation we can see if we can find a place for it adjacent to the more generic core code. We don't yet have an AITemplate-equivalent contributor repo for full kernels/device-specific implementations/etc but that's been a goal for awhile without a good example (or someone to set it up) - this may be it. The microkernels in the core repo are designed as intrinsics (like memcpy or memset) that are generic across all CPU devices, applicable to nearly all models, decoupled from transient frontend concepts/highly focused/durable, and compose with distribution and scheduling. We tend to track closer to LLVM here in terms of what makes it into intrinsics: LLVM does have an `@llvm.matrix.multiply` intrinsic but is unlikely to get `@llvm.softmax` for similar reasons to ours. That said, if this can be turned into a small sequence of instructions that are performing a more generic operation, generalized to other devices, and can apply to more than just one specific instantiation of a higher-level op then maybe we can make it a microkernel. It'd be a good example of the process of breaking down the problem and finding the right interface - ooooor~~ as mentioned this would be a good candidate for our not-yet-materialized AITemplate-equivalent out-of-core kernel library repo :)

Happy to brainstorm more (and I'm sure others would be interested) - maybe at the office hours at 4pm PST today? https://meet.google.com/ijr-xpkx-xko

**Comment by bjacob** [comment]
I just wanted to throw some :heart: as there's evidently a lot of good work that went into this. Also +1 to Scott's point about splitting large PRs.

Another way to split this discussion and unblock it might be to look into specificially what is inside Softmax.

Softmax is typically done in 3 passes:
1. Element-wise exponential.
2. Sum-reduction of the output of step 1.
3. Element-wise multiplication by the reciprocal of (i.e. division by) the output of step 2.

If softmax as a whole is deemed too coarse granularity, and we can probably agree that step 2 and step 3 are really simple things that codegen should be able to handle correctly without a ukernel, that leaves step 1, the elementwise exponential function evaluation, as the thing that one might consider a ukernel for.

SIMD element-wise math function evaluation is not easy for pure codegen to perform well on and is easier to handle in a hand-written ukernel.  It's also a finer-granularity, composable building block --- that's likely to be useful for more use cases besides Softmax. Finally, math functions such as exponential have enough arithmetic intensity (i.e. expensive enough) that the overhead of a memory-to-memory interface (as ukernels are) is not a big problem (and in the case of Softmax, the output has to be written to memory anyway).

WDYT?

**Comment by stellaraccident** [comment]
> I just wanted to throw some ❤️ as there's evidently a lot of good work that went into this. Also +1 to Scott's point about splitting large PRs.
> 
> Another way to split this discussion and unblock it might be to look into specificially what is inside Softmax.
> 
> Softmax is typically done in 3 passes:
> 
> 1. Element-wise exponential.
> 2. Sum-reduction of the output of step 1.
> 3. Element-wise multiplication by the reciprocal of (i.e. division by) the output of step 2.
> 
> If softmax as a whole is deemed too coarse granularity, and we can probably agree that step 2 and step 3 are really simple things that codegen should be able to handle correctly without a ukernel, that leaves step 1, the elementwise exponential function evaluation, as the thing that one might consider a ukernel for.
> 
> SIMD element-wise math function evaluation is not easy for pure codegen to perform well on and is easier to handle in a hand-written ukernel. It's also a finer-granularity, composable building block --- that's likely to be useful for more use cases besides Softmax. Finally, math functions such as exponential have enough arithmetic intensity (i.e. expensive enough) that the overhead of a memory-to-memory interface (as ukernels are) is not a big problem (and in the case of Softmax, the output has to be written to memory anyway).
> 
> WDYT?

+1 there's good work here that I'd like to make sure we have a path to capture as a contribution. Let's make sure to connect this with enough consulting with core team members to get this into a form that we're all positive on landing.

This will not be the last time that the granularity discussion gets evaluated differently, and softmax specifically tends to be a battleground. If we're not coming to consensus on this, let's take the time to hash it out and find a path.

**Comment by benvanik** [comment]
Nice breakdown @bjacob! I think that's a great way to start and if it ended up being just 1 (SIMD elementwise-math) then that would make for a good ukernel (reusable from lots of different lowerings) and be a good target for codegen improvements (end goal being that we in the steady-state have 0 ukernels and only use them to patch over current codegen limitations/issues). I don't fully understand why this would be hard for codegen, but having a target for investigation and evaluation is the best way to find out!

**Comment by MaheshRavishankar** [comment]
> I just wanted to throw some ❤️ as there's evidently a lot of good work that went into this. Also +1 to Scott's point about splitting large PRs.
> 
> Another way to split this discussion and unblock it might be to look into specificially what is inside Softmax.
> 
> Softmax is typically done in 3 passes:
> 
> 1. Element-wise exponential.
> 2. Sum-reduction of the output of step 1.
> 3. Element-wise multiplication by the reciprocal of (i.e. division by) the output of step 2.
> 
> If softmax as a whole is deemed too coarse granularity, and we can probably agree that step 2 and step 3 are really simple things that codegen should be able to handle correctly without a ukernel, that leaves step 1, the elementwise exponential function evaluation, as the thing that one might consider a ukernel for.
> 
> SIMD element-wise math function evaluation is not easy for pure codegen to perform well on and is easier to handle in a hand-written ukernel. It's also a finer-granularity, composable building block --- that's likely to be useful for more use cases besides Softmax. Finally, math functions such as exponential have enough arithmetic intensity (i.e. expensive enough) that the overhead of a memory-to-memory interface (as ukernels are) is not a big problem (and in the case of Softmax, the output has to be written to memory anyway).
> 
> WDYT?

Cool! This is exactly what I had in mind... Having an exponential function ukernel would really help. Maybe there is some "approximation" of exponential function that might do better. Basically if we make a distinction between data movement and instruction schedule/selection, unlike GEMM kernels, the softmax data movement is fairly straight-forward. So breaking down the performance a bit would help.
Happy to sync more, and find a way forward here without disrupting all CPU backends. Maybe break down the PR into smaller pieces.

**Comment by dcaballe** [comment]
> I just wanted to throw some ❤️ as there's evidently a lot of good work that went into this.

+1 here! Thanks a lot for the contribution and let's make sure we channel this work properly!

> SIMD element-wise math function evaluation is not easy for pure codegen to perform well on and is easier to handle in a hand-written ukernel. It's also a finer-granularity, composable building block --- that's likely to be useful for more use cases besides Softmax. Finally, math functions such as exponential have enough arithmetic intensity (i.e. expensive enough) that the overhead of a memory-to-memory interface (as ukernels are) is not a big problem (and in the case of Softmax, the output has to be written to memory anyway).
> 
> WDYT?

+1 to what has been mentioned already. I think it's key to really understand where the performance is coming from. Math related speed-ups may come from applying approximations with different trade-offs on accuracy and fp compliance. We already have a few approximations in MLIR upstream so it would be good to understand how they compare to the hand-written one (@bviyer work on some of them).

I've also seen strategies targeting softmax as a whole that introduced specific changes in the loop order/fusion of the subparts. This is also something that could be addressed by introducing specialized "structural" lowerings for the softmax itself, and then use ukernels/codegen for the resulting fine-grained ops. This is along the lines of the work that we have done on softmax so far.

**Comment by benvanik** [comment]
Cool, so sounds like we have some next steps here? This is a really great north-star and is extremely helpful in grounding the conversation - as one of the first examples plumbed through this path I'm sure it wasn't easy and having worked it all the way through with the perf numbers to show is really cool!

**Comment by MaheshRavishankar** [comment]
> Cool, so sounds like we have some next steps here? This is a really great north-star and is extremely helpful in grounding the conversation - as one of the first examples plumbed through this path I'm sure it wasn't easy and having worked it all the way through with the perf numbers to show is really cool!

+1 to that. This is some really great effort to connect all the pieces. A lot of this will be pretty useful baseline to use to figure out the final state. Thanks for pushing on this.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (ElEHsiang) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1749]: I don't want to change the behavior for other cpu targets.
Let me think about how to modify it...

File: runtime/src/iree/base/internal/cpu.c
Comment (ElEHsiang) [reply runtime/src/iree/base/internal/cpu.c:389]: Sure, I'll separate this part and send another PR.

**Comment by rednoah91** [comment]
Thanks for your feedback guys! Lowering element-wise exp as a ukernel makes sense to us because there are some trade-off implementations between accuracy and performance, and I know some HW venders make exp as a custom accelerated instruction. But we think the performance improvement is not 100% from the hand-tuned exp library part, we'll experiment just lower exp into ukernel to see how much improvement from it. We compared the code-gen with our hand-tuned kernel, and we've observed some optimizations could be considered:
(1) The `math.exp` function is invoked twice for the numerator and denominator. We can optimize this by storing the exp results in the output buffer/temp buffer, which might be a hw-independent optimization.
(2) The reduction_sum elements could be eliminated by applying vector multiple vector element-wise additions with a vector reduction addition. This might be a hw-independent optimization or maybe not. Too many elements for reduction_sum may hurt performance in some architecture.
(Please can refer to the [code](https://github.com/openxla/iree/blob/c8d25c87db98dd1fbe1a3a6f3851b25f18ea08e2/runtime/src/iree/builtins/ukernel/arch/riscv_64/softmax_riscv_64_rvv.c) for a clearer understanding.)

**Comment by dcaballe** [comment]
Great! 

> (1) The math.exp function is invoked twice for the numerator and denominator. We can optimize this by storing the exp results in the output buffer/temp buffer, which might be a hw-independent optimization.

Could you compile your benchmark with IREE and `-mlir-print-ir-after-all` and upload a gist of the output? That should help with context here.

> (2) The reduction_sum elements could be eliminated by applying vector multiple vector element-wise additions with a vector reduction addition. This might be a hw-independent optimization or maybe not. Too many elements for reduction_sum may hurt performance in some architecture.

IIUC, this is about opting for a tree-reduction vs using a vector horizontal reduction instruction, right? I think we had something to generate tree reductions somewhere but in any case it should be easy to lower to any vector reduction implementation. 

I also see a max and max reduction instruction. Make sure you use `-iree-codegen-replace-slow-min-max-ops` so that we generate the faster variants (assuming you don't care about propagating NaN values).

**Comment by ElEHsiang** [comment]
> Could you compile your benchmark with IREE and `-mlir-print-ir-after-all` and upload a gist of the output? That should help with context here.

https://gist.github.com/ElEHsiang/9ea830078b05262bddfca9c6e3fc8676
I compile the `tests/e2e/regression/softmax.mlir` with these flags. The file has near 10k lines.
There are `math.exp` in line 18771 and 17154 which are in the same function. 
```
  ../iree-build/install/bin/iree-compile \
      --iree-hal-target-backends=llvm-cpu \
  --iree-llvmcpu-target-triple=riscv64 \
  --iree-llvmcpu-target-cpu=generic-rv64 \
  --iree-llvmcpu-target-abi=lp64d \
  --iree-llvmcpu-target-cpu-features="+m,+a,+f,+d,+zvl512b,+v" \
  --riscv-v-fixed-length-vector-lmul-max=8 \
      temp/softmax.mlir \
      --mlir-print-ir-after-all \
      --mlir-print-ir-after-change \
      -o temp/softmax_aot.vmfb |& tee temp/softmax_aot.log
```
> IIUC, this is about opting for a tree-reduction vs using a vector horizontal reduction instruction, right? I think we had something to generate tree reductions somewhere but in any case it should be easy to lower to any vector reduction implementation.

I'm not sure I understand the tree-reduction but I can explain how the ukernel optimized the reduction sum in step 2.
```
1. Element-wise exponential.
2. Sum-reduction of the output of step 1.
```
The ukernel use a vector register `v_add_data` to store the sum of exponential while doing step 1. Every time the we calculate the exponential `v_exp_c`, we add it to `v_add_data`. 
And in the step 2, it simply call a reduction sum on the  `v_add_data`.

**Comment by dcaballe** [comment]
Thanks!

> There are `math.exp` in line 18771 and 17154 which are in the same function.

It looks like the softmax decomposition pass is generating this (note the two `arith.subf + math.exp` ops):

```
  %8 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1)>, affine_map<(d0, d1, d2) -> (d0, d1)>], iterator_types = ["parallel", "parallel", "reduction"]} ins(%2, %6 : tensor<12x128x128xf32>, tensor<12x128xf32>) outs(%7 : tensor<12x128xf32>) {
  ^bb0(%in: f32, %in_1: f32, %out: f32):
    %10 = arith.subf %in, %in_1 : f32
    %11 = math.exp %10 : f32
    %12 = arith.addf %11, %out : f32
    linalg.yield %12 : f32
  } -> tensor<12x128xf32>
  %9:2 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1)>, affine_map<(d0, d1, d2) -> (d0, d1)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>], iterator_types = ["parallel", "parallel", "parallel"]} ins(%2, %6, %8 : tensor<12x128x128xf32>, tensor<12x128xf32>, tensor<12x128xf32>) outs(%3, %3 : tensor<12x128x128xf32>, tensor<12x128x128xf32>) {
  ^bb0(%in: f32, %in_1: f32, %in_2: f32, %out: f32, %out_3: f32):
    %10 = arith.subf %in, %in_1 : f32
    %11 = math.exp %10 : f32
    %12 = arith.divf %11, %in_2 : f32
    linalg.yield %11, %12 : f32, f32
  } -> (tensor<12x128x128xf32>, tensor<12x128x128xf32>)
```

I guess we could make the result of the first `math.exp` and output of the generic op and forward it to the second generic op as an input. Is there a reason not to do so? The result of the second `math.exp` is also an output so I would expect no extra memory required for that?

**Comment by MaheshRavishankar** [comment]
Cc @Abhishek-Varma who has looked at this recently and might have more context on the expansion

**Comment by Abhishek-Varma** [comment]
> I guess we could make the result of the first math.exp and output of the generic op and forward it to the second generic op as an input. Is there a reason not to do so? The result of the second math.exp is also an output so I would expect no extra memory required for that?

So, Softmax decomposition in upstream LLVM [indeed does that](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp#L2573) correctly - takes in an already computed result of `math.exp` and pipes it in for division later on.

But in IREE, right after the decomposition (_within the softmax decomposition pass itself_) we are attempting to [fuse element-wise operations](https://github.com/openxla/iree/blob/main/llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/DecomposeSoftmax.cpp#L152-L168). The above IR seems to be resulting because of this fusion.

**Comment by MaheshRavishankar** [comment]
> > I guess we could make the result of the first math.exp and output of the generic op and forward it to the second generic op as an input. Is there a reason not to do so? The result of the second math.exp is also an output so I would expect no extra memory required for that?
> 
> So, Softmax decomposition in upstream LLVM [indeed does that](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp#L2573) correctly - takes in an already computed result of `math.exp` and pipes it in for division later on.
> 
> But in IREE, right after the decomposition (_within the softmax decomposition pass itself_) we are attempting to [fuse element-wise operations](https://github.com/openxla/iree/blob/main/llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/DecomposeSoftmax.cpp#L152-L168). The above IR seems to be resulting because of this fusion.

We should probably control the fusion to not fuse operations that have exp values. I think this was done because without this the code-generation does not really work. Might require this to be investigated further and fixed. I dont think anybody is actually looking at this right nowthough.

**Comment by dcaballe** [comment]
So we duplicate `arith.subf` + `math.exp` to eliminate the intermediate buffer? Not sure if there is any fusion planning stage but perhaps you can check if all ops to duplicate have only local (within the generic op) users, which is not the case for the second exp and the buffer will be generated anyways.

> I think this was done because without this the code-generation does not really work. Might require this to be investigated further and fixed. I dont think anybody is actually looking at this right nowthough.

Perhaps SiFive can give a try at disabling it and see what happens?

**Comment by ElEHsiang** [comment]
> > I think this was done because without this the code-generation does not really work. Might require this to be investigated further and fixed. I dont think anybody is actually looking at this right nowthough.
> 
> Perhaps SiFive can give a try at disabling it and see what happens?

I remove this part and there is only one `math.exp` generated. I can verify the performance and check the deference.

**Comment by benvanik** [comment]
Closing stale PR.

**Pull request closed without merging**
